File: pomade-signer/src/mailer/postmark.rs
use serde_json::json;

use crate::mailer::{Email, MailFuture, Mailer};

pub struct PostmarkMailer {
    pub client: reqwest::Client,
    pub api_token: String,
}

impl Mailer for PostmarkMailer {
    fn send(&self, from_email: &str, from_name: &str, email: Email) -> MailFuture {
        let client = self.client.clone();
        let token = self.api_token.clone();
        let from = format!("{} <{}>", from_name, from_email);
        Box::pin(async move {
            let res = client
                .post("https://api.postmarkapp.com/email")
                .header("X-Postmark-Server-Token", &token)
                .json(&json!({
                    "From": from,
                    "To": email.to,
                    "Subject": email.subject,
                    "TextBody": email.text,
                    "HtmlBody": email.html,
                }))
                .send()
                .await
                .map_err(|e| e.to_string())?;

            if res.status().is_success() {
                Ok(())
            } else {
                Err(format!("postmark error: {}", res.status()))
            }
        })
    }
}

File: pomade-signer-go/mailer/postmark.go
package mailer

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

type PostmarkMailer struct {
	Client   *http.Client
	APIToken string
}

func (m PostmarkMailer) Send(ctx context.Context, fromEmail string, fromName string, email Email) error {
	body, _ := json.Marshal(map[string]any{
		"From":     fmt.Sprintf("%s <%s>", fromName, fromEmail),
		"To":       email.To,
		"Subject":  email.Subject,
		"TextBody": email.Text,
		"HtmlBody": email.HTML,
	})
	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.postmarkapp.com/email", bytes.NewReader(body))
	req.Header.Set("X-Postmark-Server-Token", m.APIToken)
	req.Header.Set("Content-Type", "application/json")
	res, err := m.Client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode > 299 {
		return fmt.Errorf("postmark error: %s", res.Status)
	}
	return nil
}
